encoding/ssh/filexfer/buffer.go

```diff
@@ -0,0 +1,268 @@
+package filexfer
+
+import (
+	"encoding/binary"
+	"errors"
+)
+
+// Various encoding errors.
+var (
+	ErrShortPacket = errors.New("packet too short")
+	ErrLongPacket  = errors.New("packet too long")
+)
+
+// Buffer wraps up the various encoding details of the SSH format.
+//
+// Data types are encoded as per section 4 from https://tools.ietf.org/html/draft-ietf-secsh-architecture-09#page-8
+type Buffer struct {
+	b   []byte
+	off int
+}
```

Comment:
Why not

    type Buffer bytes.Buffer

? That way, we get a host of convenience methods for free.

Reply:
Because otherwise we have to read out values from the `bytes.Buffer` before we can use them. This makes the `ConsumeUintXY` functions more complicated.

Reply:
ConsumeUint32 actually gets shorter:

```go
func ConsumeUint32(b *bytes.Buffer) (uint32, error) {
	if b.Len() < 4 {
		return 0, ErrShortPacket
	}

	p := b.Next(4)
	return binary.BigEndian.Uint32(p), nil
}
```

Reply:
The Append* methods may get more expensive, though. The inliner doesn't like bytes.Buffer.WriteByte.

Reply:
Indeed. I’d also rather not have to deal with the possibly of someone Unreading anything either as well.

This gets the simplicity of everything we nee and only what we need.

Reply:
Oh, also, `type Buffer bytes.Buffer` is an opaque definition that does not import the functions of `bytes.Buffer`. In order to expose the functions of the underlying struct you have to encapsulate it: `type Buffer struct{ bytes.Buffer }`. But over-exposure of exported methods is also frustrating. Anyone could `b.Write(…)` whatever they wanted outside of spec encoding rules, and I’d really rather avoid misuse of the buffer.

Reply:
Agreed. I'd mark this resolved, but GitHub doesn't allow me to.